plugins/atomicfu/atomicfu-compiler/src/org.jetbrains.kotlinx.atomicfu.compiler/extensions/AtomicFUTransformer.kt

```diff
@@ -0,0 +1,595 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlinx.atomicfu.compiler.extensions
+
+import org.jetbrains.kotlin.backend.common.deepCopyWithVariables
+import org.jetbrains.kotlin.backend.common.extensions.IrPluginContext
+import org.jetbrains.kotlin.descriptors.DescriptorVisibilities
+import org.jetbrains.kotlin.ir.*
+import org.jetbrains.kotlin.ir.backend.js.ir.JsIrBuilder.buildValueParameter
+import org.jetbrains.kotlin.ir.expressions.impl.*
+import org.jetbrains.kotlin.ir.declarations.*
+import org.jetbrains.kotlin.ir.expressions.*
+import org.jetbrains.kotlin.ir.symbols.*
+import org.jetbrains.kotlin.ir.types.*
+import org.jetbrains.kotlin.ir.util.*
+import org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid
+import org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid
+import org.jetbrains.kotlin.ir.visitors.acceptChildrenVoid
+import org.jetbrains.kotlin.name.*
+
+private const val KOTLIN = "kotlin"
+private const val AFU_PKG = "kotlinx/atomicfu"
+private const val LOCKS = "locks"
+private const val ATOMIC_CONSTRUCTOR = "atomic"
+private const val ATOMICFU_VALUE_TYPE = """Atomic(Int|Long|Boolean|Ref)"""
+private const val ATOMIC_ARRAY_TYPE = """Atomic(Int|Long|Boolean|)Array"""
+private const val ATOMIC_ARRAY_FACTORY_FUNCTION = "atomicArrayOfNulls"
+private const val ATOMICFU_RUNTIME_FUNCTION_PREDICATE = "atomicfu_"
+private const val REENTRANT_LOCK_TYPE = "ReentrantLock"
+private const val GETTER = "atomicfu\$getter"
+private const val SETTER = "atomicfu\$setter"
+private const val GET = "get"
+private const val SET = "set"
+private const val ATOMICFU_INLINE_FUNCTION = """atomicfu_(loop|update|getAndUpdate|updateAndGet)"""
+
+private fun String.prettyStr() = replace('/', '.')
+
+class AtomicFUTransformer(override val context: IrPluginContext) : IrElementTransformerVoid(), TransformerHelpers {
+
+    private val irBuiltIns = context.irBuiltIns
+
+    private val AFU_CLASSES: Map<String, IrType> = mapOf(
+        "AtomicInt" to irBuiltIns.intType,
+        "AtomicLong" to irBuiltIns.longType,
+        "AtomicRef" to irBuiltIns.anyType,
```

Comment:
Shouldn't `AtomicRef` be nullable (`anyNType`)? According to the [declaration](https://github.com/Kotlin/kotlinx.atomicfu/blob/de4e55674c51f58cbc1f16b097d68c4ba42a5e7b/atomicfu/src/commonMain/kotlin/kotlinx/atomicfu/AtomicFU.common.kt#L123) of `AtomicRef` it's underlying type is `T : Any?`.